examples/autograd/mnist_cnn_buffer.py

```diff
@@ -0,0 +1,271 @@
+def augmentation(x, batch_size):
+    xpad = np.pad(x, [[0, 0], [0, 0], [4, 4], [4, 4]], 'symmetric')
+    for data_num in range(0, batch_size):
+        offset = np.random.randint(8, size=2)
+        x[data_num,:,:,:] = xpad[data_num, :, offset[0]: offset[0] + 28, offset[1]: offset[1] + 28]
+        if_flip = np.random.randint(2)
+        if (if_flip):
+            x[data_num, :, :, :] = x[data_num, :, :, ::-1]
+    return x
+
+def train_mnist_cnn(sgd, max_epoch, batch_size, DIST=False, data_partition=None,
+                    gpu_num=None, gpu_per_node=None, nccl_id=None, spars=0, topK=False, corr=True):
+    # Prepare training and valadiation data
+    train_x, train_y, test_x, test_y = load_dataset()
+    IMG_SIZE = 28
+    num_classes = 10
+    train_y = to_categorical(train_y, num_classes)
+    test_y = to_categorical(test_y, num_classes)
+
+    # Normalization
+    train_x = train_x / 255
+    test_x = test_x / 255
+
+    if DIST:
+        # For Distributed GPU Training
+        sgd = opt.DistOpt(sgd, nccl_id=nccl_id, gpu_num=gpu_num, gpu_per_node=gpu_per_node)
+        dev = device.create_cuda_gpu_on(sgd.rank_in_local)
+        # Dataset partition for distributed training
+        train_x, train_y = data_partition(train_x, train_y, sgd.rank_in_global, sgd.world_size)
+        test_x, test_y = data_partition(test_x, test_y, sgd.rank_in_global, sgd.world_size)
+        world_size = sgd.world_size
+    else:
+        # For Single GPU
+        print("for single GPU")
+        dev = device.create_cuda_gpu_on(0)
+        device.set_default_device(dev)
+        dev.SetRandSeed(0)
+        world_size = 1
+
+    # create model
+    print("create the model")
+    model = CNN()
+
+    # create input tensors
+    print("create input tensors")
+    tx = tensor.Tensor((batch_size, 1, IMG_SIZE, IMG_SIZE), dev, tensor.float32)
+    ty = tensor.Tensor((batch_size, num_classes), dev, tensor.int32)
+    num_train_batch = train_x.shape[0] // batch_size
+    num_test_batch = test_x.shape[0] // batch_size
+    idx = np.arange(train_x.shape[0], dtype=np.int32)
+
+    if DIST:
+        #Sychronize the initial parameters
+        autograd.training = True
+        x = np.random.randn(batch_size, 1, IMG_SIZE, IMG_SIZE).astype(np.float32)
+        y = np.zeros( shape=(batch_size, num_classes), dtype=np.int32)
+        tx.copy_from_numpy(x)
+        ty.copy_from_numpy(y)
+        out = model.forward(tx)
+        loss = autograd.softmax_cross_entropy(out, ty)
+        for p, g in autograd.backward(loss):
+            sychronize(p, sgd)
+
+    # buffer all the operations in one iteration
+    print("buffer all the operations")
+    dev.EnableGraph(True)
+    autograd.training = True
+    out = model.forward(tx)
+    loss = autograd.softmax_cross_entropy(out, ty)
+    print("buffer softmax_cross_entropy")
+    for p, g in autograd.backward(loss):
+        sgd.update(p, g)
+        # print("update sgd")
+    autograd.training = False
+    dev.EnableGraph(False)
```

Comment:
I discussed with Chris yesterday, to support distributed training I need to change some code in  [communicator.cc](../tree/dev/src/io/communicator.cc)

Reply:
> I discussed with Chris yesterday, to support distributed training I need to change some code in  communicator.cc

Yes, i.e. to buffer the all-reduce operation in communicator.cc, then it will support multi-process or MPI.
This is because the training is exactly the same as in single-gpu and multi-process (MPI), the only difference is that multi-process one has all-reduce.